core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,430 @@
+	if err1 != nil {
+		// Continuously retry from same block on any error in reorg handling.
+		return current
+	}
+	// If we did detect a reorg, we'll have a new block number to start from (LCA+1)
+	// so let's resume polling from there.
+	if reorgDetected {
+		current = newPollBlock
+	}
+
+	// Backfill finalized blocks if we can for performance.
+	// E.g. 1<-2<-3(current)<-4<-5<-6<-7(latest), finality is 2. So 3,4,5 can be batched.
+	// start = current = 3, end = latest - finality = 7-2 = 5 (inclusive range).
+	if (latest - current) >= lp.finalityDepth {
+		lp.lggr.Infow("Backfilling logs", "start", current, "end", latest-lp.finalityDepth)
+		current = lp.backfill(ctx, current, latest-lp.finalityDepth)
+	}
+
+	for current <= latest {
+		// Same reorg detection on unfinalized blocks.
+		// Get current block
+		block, err2 := lp.ec.BlockByNumber(ctx, big.NewInt(current))
+		if err2 != nil {
+			lp.lggr.Warnw("Unable to get block", "err", err1, "current", current)
+			return current
+		}
+		reorgDetected, newPollBlock, err2 := lp.maybeHandleReorg(block, current, latest)
+		if err2 != nil {
+			return current
+		}
+		if reorgDetected {
+			current = newPollBlock
+			continue
+		}
+
+		h := block.Hash()
+		logs, err2 := lp.ec.FilterLogs(ctx, ethereum.FilterQuery{
+			BlockHash: &h,
+			Addresses: lp.filterAddresses(),
+			Topics:    lp.filterTopics(),
+		})
+		if err2 != nil {
+			lp.lggr.Warnw("Unable query for logs, retrying", "err", err2, "block", block.Number())
+			return current
+		}
+		lp.lggr.Infow("Unfinalized log query", "logs", len(logs), "current", current)
+		err2 = lp.orm.q.Transaction(func(q pg.Queryer) error {
+			if err3 := lp.orm.InsertBlock(block.Hash(), block.Number().Int64()); err3 != nil {
+				return err3
+			}
+			if len(logs) == 0 {
+				return nil
+			}
+			return lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs))
+		})
+		if err2 != nil {
+			// If we're unable to insert, don't increment current and just retry
+			lp.lggr.Warnw("Unable to save logs, retrying", "err", err2, "block", block.Number())
+			return current
+		}
+		current++
+	}
+	return current
+}
+
+func (lp *LogPoller) findLCA(h common.Hash, mismatchStart int64) (int64, error) {
+	// Find the first place where our chain and their chain have the same block,
+	// that block number is the LCA.
+	block, err := lp.ec.BlockByHash(context.Background(), h)
+	if err != nil {
+		return 0, err
+	}
+	number := block.Number().Int64()
+	if (mismatchStart - number) > lp.finalityDepth {
+		lp.lggr.Criticalw("Reorg greater than finality depth detected", "depth", (mismatchStart - number), "finality", lp.finalityDepth)
+		return 0, errors.New("reorg greater than finality depth")
+	}
+	ourBlockHash, err := lp.orm.SelectBlockByNumber(block.Number().Int64())
+	if err != nil {
+		return 0, err
+	}
+	if !bytes.Equal(block.Hash().Bytes(), ourBlockHash.BlockHash.Bytes()) {
+		return lp.findLCA(block.ParentHash(), mismatchStart)
```

Comment:
nit: could convert this to a for loop

Reply:
any issues you see as is? Recursion feels intuitive here

Reply:
wouldn't bet on go's tail call support (couldn't find material that it's actually optimized and won't blow up the stack) honestly and would prefer that very large reorgs can't make us crash. we might have differences in orders of magnitude in the expected reorg sizes between different chains

Reply:
yeah fair enough, looks like they don't optimize